### ISSUE TITLE
Clarification on the returned shape for the gradients

### ISSUE THREAD

**Issue by Stochastic13** [philipperemy/keract#30]
In regards to the output of `get_gradients_of_activations` vs `get_gradients_of_trainable_weights`:

I have a `Conv2D` layer with `16` `(5,5)` filters and `padding = 'same'`, which accepts input of dimension `(batch_size, 96, 96, 10)`. 
Let's say I am using a batch of 2 inputs (`x.shape = (2, 96, 96, 10)`) vs a batch of a single input (`x.shape = (1, 96, 96, 10)`).
So the shapes returned by the three functions are as such:

    get_activations: (2, 96, 96, 16), (1, 96, 96, 16)
    get_gradients_of_activations: (2, 96, 96, 16), (1, 96, 96, 16)
    get_gradients_of_trainable_weights - kernel: (5, 5, 10, 16), (5, 5, 10, 16)
    get_gradients_of_trainable_weights - bias: (16,), (16,)

So, am I right in assuming that `trainable_weights` gradients are averaged over all the training examples, but the `activation` gradients are individually returned for each input sample?

Also, `activation` gradients are not the gradients of the activation function, are they? Because then they'd be used to backpropagate and be included in the gradients for `trainable_weights`?

**Comment by philipperemy** [comment]
```
from __future__ import print_function

import numpy as np
from keras.layers import Conv2D
from keras.layers import Dense, Flatten
from keras.models import Sequential

if __name__ == '__main__':
    model = Sequential()
    model.add(Conv2D(16, kernel_size=(5, 5), padding='same', input_shape=(96, 96, 10), name='my_conv'))
    model.add(Flatten())
    model.add(Dense(128, activation='relu'))
    model.add(Dense(1, activation='sigmoid'))

    model.compile(loss='mse',
                  optimizer='adam',
                  metrics=['accuracy'])

    x1 = np.random.uniform(size=(1, 96, 96, 10))
    x2 = np.random.uniform(size=(2, 96, 96, 10))

    y1 = np.random.uniform(size=(1,))
    y2 = np.random.uniform(size=(2,))

    import keract


    def run(x, y):
        bs = len(x)
        print(bs)
        print('get_activations - my_conv/BiasAdd:0',
              keract.get_activations(model, x)['my_conv/BiasAdd:0'].shape)

        print('get_gradients_of_activations - my_conv/BiasAdd:0',
              keract.get_gradients_of_activations(model, x, y)['my_conv/BiasAdd:0'].shape)

        print('get_gradients_of_trainable_weights - my_conv/kernel:0',
              keract.get_gradients_of_trainable_weights(model, x, y)['my_conv/kernel:0'].shape)

        print('get_gradients_of_trainable_weights - my_conv/bias:0',
              keract.get_gradients_of_trainable_weights(model, x, y)['my_conv/bias:0'].shape)


    print('*' * 80)
    run(x1, y1)
    print('*' * 80)
    run(x2, y2)
```


```
1
get_gradients_of_activations - my_conv/BiasAdd:0 (1, 96, 96, 16)
get_gradients_of_trainable_weights - my_conv/kernel:0 (5, 5, 10, 16)
get_gradients_of_trainable_weights - my_conv/bias:0 (16,)
********************************************************************************
2
get_activations - my_conv/BiasAdd:0 (2, 96, 96, 16)
get_gradients_of_activations - my_conv/BiasAdd:0 (2, 96, 96, 16)
get_gradients_of_trainable_weights - my_conv/kernel:0 (5, 5, 10, 16)
get_gradients_of_trainable_weights - my_conv/bias:0 (16,)
```

**Comment by philipperemy** [comment]
If this explanation is good enough, I will add it in the README :)

**Comment by Stochastic13** [comment]
Thanks for the explanation! It is good enough for me. :)

PS: Was thinking of a way to visualize these `gradients_for_trainable_weights` . One idea is to present `out_channels` number of rows, each being made from unraveled `in_channel` number of images of dimensions `filter_w , filter_h` . Personally, such a visualization would have come in handy to debug vanishing gradients with `relu` . I can send a PR if this idea seems useful to you.

**Comment by philipperemy** [comment]
@Stochastic13 looks good to me. I'd be happy to review a PR on this. Will def be useful :)
